Refactor/travel schedule structure change

ErranderLee opened the pull request:
### 🔨 작업 내용
- travel, schedule 관련 구조 변경
### 📐 구현한 내용
- travel, travelDate -> 1:N, travelDate, schedule -> 1:N 으로 구조 변경
  - travelDate Primary Key로 date와 travelId를 포함한 composite key를 사용
  - 구조 변경에 따른 cumstom error 추가
  - 구조 변경에 따른 api 수정
  - 구조 변경에 따른 test 수정
  - travelDate create, delete api 구현
### 🚧 논의 사항
-

ErranderLee committed:
refactor:
 - travel, schedule entity reconstruct
   - add travelDate entity
   - add travelDateOrder to travel
   - add ScheduleOrder to travelDate
   - Modify Relation Mapping

ErranderLee committed:
refactor:
 - modify TravelResponseDto
 - make SimpleTravelDateResponseDto
 - remove api that get schedules by travelId
 - modify update schedule order api
 - add custom error(DATE_NOT_FOUND)
 - move find schedules with place method from TravelRepository to TravelDateRepository
 - modify getTravelId method
   - get all related dates with all related schedule with place

ErranderLee committed:
refactor:
 - create TravelDate

ErranderLee committed:
refactor:
 - remove startDate and endDate in Travel
 - modify Dtos following above issue
 - make create TravelDate api
 - modify ScheduleService and TravelService

ErranderLee committed:
refactor:
 - remove autoincrement id from TravelDate
 - add composite key to TravelDate

ErranderLee committed:
refactor:
 - contains travelId and Date in ScheduleResponseDto

ErranderLee committed:
fix:
 - resolve error.
   - query exception occured by invalid object searching

ErranderLee committed:
refactor:
 - modify test and src and fix error

ErranderLee committed:
refactor: code refactor and fix
 - modify DTOs
 - fix composite key problems
 - fix custom jpql
 - getSchedules by travelDate(travelId and date)

fix: code fix
 - fix test

ErranderLee committed:
fix:
 - modify test and repository method
 - add test to test change schedule order

park0503 commented on src/main/java/com/ajou/travely/exception/ErrorCode.java:
```diff
@@ -10,6 +10,7 @@ public enum ErrorCode {
     SCHEDULE_NOT_FOUND(400, "SCHEDULE RECORD IS NOT FOUND"),
     TRAVEL_NOT_FOUND(400, "TRAVEL RECORD IS NOT FOUND"),
     FRIEND_NOT_FOUND(400, "FRIEND RECORD IS NOT FOUND"),
+    DATE_NOT_FOUND(400, "DATE RECORD IS NOT FOUND"),
```
TRAVEL_DATE로 바꿔도 될 것 같아요!

ErranderLee replied:
반영했습니다~~

park0503 commented on src/main/java/com/ajou/travely/service/TravelService.java:
```diff
@@ -270,6 +293,14 @@ private Invitation checkInvitationRecord(UUID code, String email) {
+        return checkRecord(
+                travelDateRepository.findById(travelDateId),
+                "해당 ID의 Date가 존재하지 않습니다.",
+                ErrorCode.DATE_NOT_FOUND
```
여기도 Date라기 보다는 Travel Date와 같이 명확한 Type을 명시해 주면 추후 혼선을 예방할 수 있을 것 같단 생각이 듭니다!

ErranderLee replied:
반영했습니다!

park0503 commented on src/main/java/com/ajou/travely/controller/travel/TravelController.java:
```diff
@@ -182,8 +183,8 @@ public ResponseEntity<Void> deleteCost(@PathVariable Long travelId,
     @PostMapping("/{travelId}/travelDates")
-    public ResponseEntity<LocalDate> createTravelDate(@PathVariable Long travelId,
-                                                      @Valid @RequestBody TravelDateCreateRequestDto requestDto) {
+    public ResponseEntity<TravelDate> createTravelDate(@PathVariable Long travelId,
+                                                       @Valid @RequestBody TravelDateCreateRequestDto requestDto) {
```
Entity를 직접 보내는 것 보단 dto를 따로 만드는 것이 어떨까 싶습니다!

ErranderLee replied:
반영했습니다!

park0503 commented on src/main/java/com/ajou/travely/service/TravelService.java:
```diff
@@ -214,24 +208,20 @@ public List<SimpleCostResponseDto> getCostsByTravelId(Long travelId) {
+            .stream()
+            .map(SimpleScheduleResponseDto::new)
+            .collect(Collectors.toList());
+    }
```
변경된 구조에 맞게 메소드명을 변경해도 좋을 것 같습니다.

ErranderLee replied:
반영했습니다!

ErranderLee committed:
fix:
 - resolve merge conflicts
 - modify create schedule api(path) and ScheduleCreateRequestDto

ErranderLee committed:
fix:
 - remove PostRepositoryTest

park0503 commented on src/main/java/com/ajou/travely/service/TravelService.java:
```diff
@@ -272,16 +255,12 @@ private List<Schedule> sortSchedule(TravelDate travelDate) {
 
     @Transactional
-    public LocalDate createTravelDate(Long travelId, TravelDateCreateRequestDto requestDto) {
+    public TravelDate createTravelDate(Long travelId, TravelDateCreateRequestDto requestDto) {
```
해당 메소드의 반환 값은 controller에 반환될 값인 것 같은데, Entity가 한 Layer를 벗어나는 것은 영속성 컨텍스트 관리 측면에서 어려움이 생길 수 있다고 강의 때 들은적이 있습니다. 적절한 DTO를 만들어 반환해 주면 좋을 것 같습니다.

ErranderLee replied:
좋은 정보 알려주셔서 감사합니다! 반영하겠습니다.

park0503 reviewed (commented):
정말 긴 코드 고생 많으셨습니다. 쉽지 않은 작업이었을텐데 잘 끝내주셔서 감사합니다.
몇부분 관련해서 리뷰 남겨놨습니다. 확인 부탁드릴게요~

ErranderLee committed:
fix:
 - modify ErrorCode and Message about travelDate
 - make TravelDateCreateResponseDto not to return Entity
 - modify test

ErranderLee committed:
fix:
 - modify method with proper name
 - modify test about changes

park0503 reviewed (commented):
😎👍

Sophoca reviewed (commented):
고생하셨습니다. 머지하겠습니다.
